test/unit/integrated_function_test.C

```diff
@@ -0,0 +1,309 @@
+        {
+          libMesh::Real costheta = std::cos(theta[t]);
+          libMesh::Real sintheta = std::sin(theta[t]);
+          libMesh::Real L = 3.0/costheta;
+
+          std::vector<std::string> functions;
+          std::vector<libMesh::Real> calc_answers;
+          std::vector<bool> converged;
+
+          // trig
+          functions.push_back("sin(x)+cos(y)");
+
+          if (theta[t] == 0.0)
+            calc_answers.push_back( L-std::cos(L)+1.0 );
+          else
+            calc_answers.push_back( ( (std::sin(L*sintheta))/sintheta ) - ( (std::cos(L*costheta)-1.0)/costheta ) );
+
+          converged.push_back(false);
+
+          // exponential
+          functions.push_back("exp(x)");
+          calc_answers.push_back( (std::exp(3.0) - 1.0) / costheta);
+          converged.push_back(false);
+
+          std::vector<std::vector<libMesh::Real> > errors(functions.size());
+          std::vector<std::vector<libMesh::Real> > h_vals(functions.size());
+
+          GRINS::MultiphysicsSystem* system = _sim->get_multiphysics_system();
+          GRINS::SharedPtr<libMesh::EquationSystems> es = _sim->get_equation_system();
+
+          // build rayfire
+          GRINS::SharedPtr<GRINS::RayfireMesh> rayfire = new GRINS::RayfireMesh(origin,theta[t]);
+
+          CPPUNIT_ASSERT_EQUAL(functions.size(), calc_answers.size());
+
+          // and now the CompositeQoI to do the evalutaion
+          GRINS::CompositeQoI comp_qoi;
+
+          for (unsigned int i=0; i<functions.size(); i++)
+            {
+              GRINS::IntegratedFunction<libMesh::FunctionBase<libMesh::Real> > integ_func((unsigned int)3,new libMesh::ParsedFunction<libMesh::Real>(functions[i]),rayfire,"integrated_function");
+              comp_qoi.add_qoi(integ_func);
+            }
+
+          comp_qoi.init(*_input,*system);
+          system->attach_qoi(&comp_qoi);
+
+          libMesh::MeshRefinement mr(system->get_mesh());
+
+          unsigned int num_converged = 0;
+          unsigned int iter = 0;
+
+          // limit iterations to prevent excessive runtime
+          unsigned int max_iter = 7;
+          do
+            {
+              std::vector<libMesh::dof_id_type> elems_in_rayfire;
+              rayfire->elem_ids_in_rayfire(elems_in_rayfire);
+
+              system->assemble_qoi();
+
+              libMesh::Real h = -1.0;
+              for (unsigned int i=0; i<elems_in_rayfire.size(); i++)
+                {
+                  libMesh::Real l = (rayfire->map_to_rayfire_elem(elems_in_rayfire[i]))->length(0,1);
+                  if (l>h)
+                      h=l;
+                }
+
+              CPPUNIT_ASSERT(h > 0.0);
+
+              for (unsigned int i=0; i<calc_answers.size(); i++)
+                {
+                  if (!converged[i])
+                    {
+                      libMesh::Real err = std::abs( _sim->get_qoi_value(i) - calc_answers[i] );
+                      if (err < 2.5e-9)
```

Comment:
They both take about the same number of iterations now. I had to pick the tolerance carefully so that the exponential didn't need another refinement, since that more than doubled the runtime. I can have a separate tolerance for each function if you want